Sinkhorn Divergence

davibarreira opened the pull request:
Implemented the function `sinkhorn_divergence` which is an unbiased version of `sinkhorn` and that is also a metric in the space of probability spaces. This function is similar to the `ot.bregman.empirical_sinkhorn_divergence` from `POT.py`.

The tests required the use of `PyCall`, because `ot.bregman.empirical_sinkhorn_divergence` is not supported on `PythonOT.jl`.

davibarreira committed:
Initianning sikhorn divergence

davibarreira committed:
Merge branch 'master' of https://github.com/JuliaOptimalTransport/OptimalTransport.jl into sinkhorndivergence

davibarreira committed:
Sinkhorn divergence implemented

davibarreira committed:
Added PyCall to test dependencies

davibarreira committed:
Added tests for sinkhorn divergence

davibarreira committed:
Added Sinkhorn Divergence to docs

devmotion commented:
> The tests required the use of PyCall, because ot.bregman.empirical_sinkhorn_divergence is not supported on PythonOT.jl

It would be good to add it to PythonOT.

davibarreira commented:
> > The tests required the use of PyCall, because ot.bregman.empirical_sinkhorn_divergence is not supported on PythonOT.jl
> 
> It would be good to add it to PythonOT.

I'll see how this can be done,

davibarreira commented:
Submitted a PR to the `PythonOT.jl` package. Once the PR goes through there, I'll update this PR here in order to remove the `PyCall` dependency on the  test.

zsteve commented on src/OptimalTransport.jl:
````diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+
+The Sinkhorn Divergence is computed as:
+```math
+S_{c,ε}(μ,ν) := OT_{c,ε}(μ,ν) - \\frac{1}{2}(OT_{c,ε}(μ,μ) + OT_{c,ε}(ν,ν)),
````
Could we use `\\operatorname{OT}` here instead?

zsteve replied:
Same for `\\operatorname{S}`

davibarreira replied:
Yeah, better to add the `\\operatorname`. Will do.

zsteve commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+        regularization=regularization,
+        kwargs...,
+    )
+    return max(0, OTμν - 0.5 * (OTμμ + OTνν))
```
Is there a reason we take `max(0, ...)` here?

davibarreira replied:
To guarantee that the value is larger or equal to 0. The same is also present in `POT.py`.

zsteve replied:
Makes sense. It doesn't come from the 'math' though, I assume any negative values would be due to numerical issues (i.e. any negative values would be very small?)

davibarreira replied:
Correct. In theory, this is a proper metric, so it would never be negative.

zsteve commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+See also: [`sinkhorn2`](@ref)
+"""
+function sinkhorn_divergence(
+    c,
```
Could we stick to the calling convention of `sinkhorn` here, i.e. `(\mu, \nu, C, \varepsilon)` in that order.

davibarreira replied:
I thought we were going to revert all functions to `(c,\mu,\nu ...)`, like the implementations of `ot_cost` and `ot_plan`. Also, I think we should use lowercase `c` for the cost function, and uppercase `C` for cost matrix.

zsteve replied:
Ah I see. Makes sense, which PR is this? Sorry I've not kept up with all of them :P

davibarreira replied:
Not your fault at all! I thought we had settle on it in the issue #63 , but now I see it's not there. So probably me and @devmotion talked about in one of the Multivariate Normals or 1D implementations (which have almost hundreds of comments). Things are moving so fast in this package now, that I actually don't know where it is.
I'll comment in issue #63, so we can make a decision. If I remember correctly, it was actually a suggestion of @devmotion, based on the fact that one usually declares arguments that are functions in the beginning (I'm guessing this is a Julia standard or something).

devmotion replied:
Yes, this is a standard and mentioned in the Julia documentation about ordering of function arguments since it allows one to use the `do` syntax
```julia
map(rand(20)) do x
    return x^2
end
```
instead of
```julia
map(x -> x^2, rand(20))
```
which is very convenient for longer and more complicated functions.

devmotion commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+See also: [`sinkhorn2`](@ref)
+"""
+function sinkhorn_divergence(
+    c,
```
Could we implement this with the cost matrices as inputs instead? I think this would have multiple advantages:
- it would allow to reuse precomputed cost matrices, here and in other functions
- it would be guaranteed to work on the GPU (BTW GPU tests should be added if possible) whereas the default implementation of `pairwise` probably doesn't for most distances and custom functions (IIRC it uses incorrect containers and indexing which should be avoided on GPUs and is often disabled by users)

devmotion replied:
Additionally, this would still allow to define the function implemented by just forwarding it to the one with cost matrices that are computed from the `DiscreteNonParametric` and the cost function.

devmotion replied:
BTW it's also quite restrictive to only allow univariate messures here it seems.

davibarreira replied:
Yeah, the univariate was a mistake. I forgot that `DiscreteNonParametric` was only for the univariate case. I could implement it with a matrix cost, but it would actually require 3 matrices, so the it would be quite "ugly". It would be something like:
`sinkhorn_divergence(mu,nu,C1,C2,C3,eps)`. So I think we should stick with the cost function... I don't know much about programming for GPU, but, don't you think we could adapt this somehow?

davibarreira replied:
I could implement another `sinkhorn_divergence` that would take the cost matrices as argument. And would could perhaps indicate that such version should be used in case the user is interested in using GPU.

devmotion replied:
You can just define both and forward the call from the one with the cost function to the other:

> Additionally, this would still allow to define the function implemented by just forwarding it to the one with cost matrices that are computed from the DiscreteNonParametric and the cost function.

devmotion replied:
>  such version should be used in case the user is interested in using GPU.

It could be useful without GPUs as well, e.g., if you evaluate `sinkhorn`, `sinkhorn2` or other other regularized OT distances with the same cost matrices.

In fact, it could also be useful to allow to specify pre-computed plans for the different `sinkhorn2` calls.

davibarreira replied:
Convinced.

devmotion commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+    μ::DiscreteNonParametric,
+    ν::DiscreteNonParametric,
+    ε;
+    regularization=false,
```
This can be removed since it's the default in `sinkhorn2`:
```suggestion
```

devmotion commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+        ν.p,
+        pairwise(c, μ.support, ν.support),
+        ε;
+        regularization=regularization,
```
```suggestion
```

devmotion commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+        μ.p,
+        pairwise(c, μ.support; symmetric=true),
+        ε;
+        regularization=regularization,
```
```suggestion
```

devmotion commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+        ν.p,
+        pairwise(c, ν.support; symmetric=true),
+        ε;
+        regularization=regularization,
```
```suggestion
```

devmotion commented on src/OptimalTransport.jl:
```diff
@@ -785,4 +786,68 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+        regularization=regularization,
+        kwargs...,
+    )
+    return max(0, OTμν - 0.5 * (OTμμ + OTνν))
```
We should avoid unwanted promotions (e.g. when working with `Float32` on GPUs):
```suggestion
    return max(0, OTμν - (OTμμ + OTνν) / 2)
```

Intuitively I would have thought that the divergence can be negative if the regularization terms are included (BTW does one actually want to include them)?

davibarreira replied:
Yeah, if the regularization terms are added, it could be negative, so you are correct that we should not allow the `regularization` keyword. For the Sinkhorn Divergence, one should probably never include the regularization terms, because the goal here is to turn the entropic ot cost into a "proper" metric in the space of probability distributions. Besides, Feydy (in the paper I cited), proved some neat properties of this metric, such as the fact that it also metrizes weak convergence.
TL;DR: I don't think one would want to ever include the regularization terms in the Sinkhorn Divergence.

For `sinkhorn2`, I saw one paper where the author suggested that the use of the cost plus regularization to behave better than when one removes it. Although, it can be negative, so it loses some of the interpretability.

devmotion replied:
Thanks for the detailed explanation! This confirms my intuition, so let's just remove the `regularization` keyword argument. Maybe then we should even check that `kwargs...` does not contain a keyword argument `regularization` and otherwise either throw an error or (maybe better) drop the `regularization` keyword argument and display a `@warn`ing. Maybe something like
```julia
....; regularization=nothing, kwargs...)
    if regularization !== nothing
        @warn "`sinkhorn_divergence` does not support the `regularization` keyword argument"
    end
```

devmotion commented on test/entropic.jl:
```diff
@@ -213,4 +217,26 @@ Random.seed!(100)
+        @testset "example" begin
+            # create distributions 
+            N = 100
+            μ = DiscreteNonParametric(rand(N), ones(N) / N)
```
Can we use random histograms?

davibarreira replied:
Yeah. I'll update.

davibarreira committed:
Creating FiniteDiscreteMeasure struct

davibarreira committed:
Modifications:
- Created the struct FiniteDiscreteMeasure,
- Implemented two versions of sinkhorn_divergence,
- Disabled the use of regularization on sinkhorn_divergence,
- Fixed docstring with suggestions.

davibarreira committed:
FixedDiscreteMeasure normalizes the weights to sum 1

davibarreira committed:
FixedDiscreteMeasure checks if probabilities are positive

davibarreira committed:
Created tests for FiniteDiscreteMeasure

davibarreira committed:
Added tests for sinkhorn divergence and finite discrete measure

davibarreira committed:
Fixed the code for creating cost matrices in the sinkhorn_divergence

davibarreira committed:
Added costmatrix.jl to tests

davibarreira committed:
Fixed docstring for costmatrix

davibarreira committed:
Fixed errors in the tests

davibarreira committed:
Minor fixes in the tests

davibarreira commented:
So, this current push in the PR is actually too much code for a single PR. I intend to break it down in other PR's, but, I thought it would be good to submit here the whole thing so you guys can take a look on how these new functions tie themselves to the rest of the package.
So, besides the `sinkhorn_divergence` function, I actually implemented the struct `FiniteDiscreteMeasure` and the `cost_matrix` function. Each one I'll separate in a different PR, in case we decide that they are appropriate for the package.

The `FiniteDiscreteMeasure` is a struct is a way that we can define our empirical measures and consistently write functions such as `sinkhorn2(c, mu, nu)` or `sinkhorn2(C, mu, nu)`. Right now, the package sometimes refers `mu` and `nu` only as the support, and sometimes as actual measures. So the new struct would clear things.

The `cost_matrix` is just a helper function to construct cost matrices easily. This allows us to create other versions of our functions such as `sinkhorn2` with a cost function instead of a cost matrix (similar to what I have done with sinkhorn_divergence).

@devmotion  and @zsteve , what do you think of these two new additions? Should I submit a PR for each one?

devmotion commented:
I think it's a good idea to move these two additions to separate PRs, then it's easier to discuss and polish them.

I also think that probably a `FiniteDiscreteMeasure` is useful, I have some additional suggestions but it's easier to take this in the corresponding PR.

I am a bit less convinced by `cost_matrix`, it seems to introduce some type instabilities but probably this could be fixed. I don't think it should be exported and become part of the official API though, but also here it's probably best to discuss this in a separate PR :slightly_smiling_face:

davibarreira committed:
Fixed tests and merged

davibarreira commented:
The error in the check seems to be unrelated to the PR.

devmotion commented on Project.toml:
```diff
@@ -35,6 +35,7 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 SafeTestsets = "1bc83da4-3b8d-516f-aca4-4fe02f6d838f"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 Tulip = "6dd1b50a-3aae-11e9-10b5-ef983d2400fa"
+PyCall = "438e738f-606a-5dbb-bf0a-cddfbfd45ab0"
```
Not needed anymore I guess?
```suggestion
```

devmotion commented on src/OptimalTransport.jl:
```diff
@@ -4,6 +4,7 @@
 
 module OptimalTransport
 
+using LinearAlgebra: AbstractMatrix
```
```suggestion
```

devmotion commented on src/entropic/sinkhorn.jl:
```diff
@@ -504,3 +504,85 @@ function sinkhorn_barycenter(μ, C, ε, w; tol=1e-9, check_marginal_step=10, max
+"""
+    sinkhorn_divergence(
+        c,
+        μ::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
+        ν::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
```
It's not really what we support, maybe just omit the type or add two docstrings (one basically referring to the other)?

devmotion commented on src/entropic/sinkhorn.jl:
```diff
@@ -504,3 +504,85 @@ function sinkhorn_barycenter(μ, C, ε, w; tol=1e-9, check_marginal_step=10, max
+        c,
+        μ::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
+        ν::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
+        ε; regularization=false, plan=nothing, kwargs...
```
It would be useful if we would restructure `sinkhorn` and introduce the different algorithms, as proposed in the other PR. Then users could also choose the stabilized algorithm or epsilon scaling here.

devmotion replied:
One should not include the regularization keyword argument since it is not supported.

```suggestion
        ε; plan=nothing, kwargs...
```

devmotion commented on src/entropic/sinkhorn.jl:
```diff
@@ -504,3 +504,85 @@ function sinkhorn_barycenter(μ, C, ε, w; tol=1e-9, check_marginal_step=10, max
+"""
+    sinkhorn_divergence(
+        Cμν, Cμμ, Cνν,
+        μ::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
+        ν::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
```
Same here.

devmotion commented on src/entropic/sinkhorn.jl:
```diff
@@ -504,3 +504,85 @@ function sinkhorn_barycenter(μ, C, ε, w; tol=1e-9, check_marginal_step=10, max
+        Cμν, Cμμ, Cνν,
+        μ::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
+        ν::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
+        ε; regularization=false, plan=nothing, kwargs...
```
```suggestion
        ε; plan=nothing, kwargs...
```

devmotion commented on test/entropic/sinkhorn.jl:
```diff
@@ -195,4 +196,93 @@ Random.seed!(100)
+            μ = OptimalTransport.discretemeasure(μsupp, μprobs)
+            ν = OptimalTransport.discretemeasure(νsupp)
+
+            for (ε, metrics) in Iterators.product(
+                [0.1, 1.0, 10.0],
+                [
+                    (sqeuclidean, SqEuclidean()),
+                    (euclidean, Euclidean()),
+                    (totalvariation, TotalVariation()),
+                ],
+            )
```
This is a bit uncommon, usually one would just use two `for` loops here. One can even use the short notation
```julia
for x in xs, y in ys
```

devmotion commented on test/entropic/sinkhorn.jl:
```diff
@@ -195,4 +196,93 @@ Random.seed!(100)
+            μ = OptimalTransport.discretemeasure(μsupp, μprobs)
+            ν = OptimalTransport.discretemeasure(νsupp)
+
+            for (ε, metrics) in Iterators.product(
```
Same here.

devmotion commented on test/finitediscretemeasure.jl:
```diff
@@ -0,0 +1,55 @@
+using Distributions: DiscreteNonParametric
```
This file seems unrelated to the PR?

devmotion commented on test/runtests.jl:
```diff
@@ -1,3 +1,4 @@
+using LinearAlgebra: symmetric
```
I guess not needed?

```suggestion
```

davibarreira committed:
Update Project.toml

Co-authored-by: David Widmann <devmotion@users.noreply.github.com>

davibarreira committed:
Update src/OptimalTransport.jl

Co-authored-by: David Widmann <devmotion@users.noreply.github.com>

davibarreira committed:
Update test/runtests.jl

Co-authored-by: David Widmann <devmotion@users.noreply.github.com>

davibarreira committed:
Update src/entropic/sinkhorn.jl

Co-authored-by: David Widmann <devmotion@users.noreply.github.com>

davibarreira committed:
Update src/entropic/sinkhorn.jl

Co-authored-by: David Widmann <devmotion@users.noreply.github.com>

davibarreira committed:
removed incorrect file

davibarreira committed:
Simplified loop format in the test

davibarreira committed:
Adjusted function to the new Sinkhorn format

davibarreira committed:
Fixed format

davibarreira commented:
I think this one is finally ready to go.

davibarreira commented:
@devmotion or @zsteve, whenever you guys can, please review this PR.
